### PR TITLE
add rate limit to asset inventory

### PR DESCRIPTION
**Pull request by orouz**
### Summary of your changes

1. adds rate limiting to the asset inventory client. our current usage is only for `ListAssets`
2. retries requests made by `ListAssets` client whenever they failed due to rate limiting
3. adds a cache to the policies fetcher (using a new cache utility, as we already have 3 caches in this file)

for `ListAssets`, we only use the per-project [quota](https://cloud.google.com/asset-inventory/docs/quota), which is `100 per minute per consumer project`. we do this because:
1. the consumer project (the one consuming the quota) is set by the user when they run `gcloud config set project <project_id>` before deploying the agent. (verified with `gcloud config get billing/quota_project`). 
2. in both our use cases: `single-account` and `organization-account`, we always use a single quota project. we never re-define it for the user
4. in anyway, the org quotas are `800 per minute per org` and `650,000 per day per org`. so the per-project quota is more restrictive than both of these, meaning we shouldn't exceed those either.
5. we don't account for multiple cloudbeat instances running together without being synced on the quotas each of them consume (this is assumed to be an unlikely edge case we accept for now)
 
### Screenshot/Data

<details>
<summary>test script</summary>

```go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	asset "cloud.google.com/go/asset/apiv1"
	"cloud.google.com/go/asset/apiv1/assetpb"
	"github.com/googleapis/gax-go"
	"golang.org/x/time/rate"
	"google.golang.org/api/iterator"
	"google.golang.org/api/option"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
)

type Iterator interface {
	Next() (*assetpb.Asset, error)
}

var RetryOnResourceExhausted = gax.WithRetry(func() gax.Retryer {
	return gax.OnCodes([]codes.Code{codes.ResourceExhausted}, gax.Backoff{
		Initial:    1 * time.Second,
		Max:        10 * time.Second,
		Multiplier: 1.2,
	})
})

type AssetsInventoryRateLimiter struct {
	methods map[string]*rate.Limiter
}

const projectQuota = 100

func NewAssetsInventoryRateLimiter() *AssetsInventoryRateLimiter {
	return &AssetsInventoryRateLimiter{
		methods: map[string]*rate.Limiter{
			"/google.cloud.asset.v1.AssetService/ListAssets": rate.NewLimiter(rate.Every(time.Minute/projectQuota), 1),
		},
	}
}

func (rl *AssetsInventoryRateLimiter) GetInterceptorDialOption() grpc.DialOption {
	return grpc.WithUnaryInterceptor(func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
		limiter := rl.methods[method]
		if limiter != nil {
			limiter.Wait(ctx)
		}
		return invoker(ctx, method, req, reply, cc, opts...)
	})
}

func main() {
	ctx := context.Background()
	limiter := NewAssetsInventoryRateLimiter()
	clientA, err := asset.NewClient(ctx, option.WithGRPCDialOption(limiter.GetInterceptorDialOption()))

	if err != nil {
		log.Fatalf("failed to create client: %v", err)
	}

	var totalGot int
	var totalLost int
	start := time.Now()
	// simulate 10x the project per-minute quota by requesting some assets multiple times
	var assets []*assetpb.Asset
	for i := 0; i < projectQuota*10; i++ {
		log.Printf("Iteration: %d \n", i)
		resp := getAllAssets(clientA.ListAssets(ctx, &assetpb.ListAssetsRequest{
			Parent:      fmt.Sprintf("organizations/%s", "693506308612"),
			AssetTypes:  []string{"logging.googleapis.com/LogBucket"},
			ContentType: assetpb.ContentType_RESOURCE,
		}, RetryOnResourceExhausted))
		if resp == nil {
			totalLost++
		} else {
			totalGot++
			assets = append(assets, resp...)
		}
	}
	end := time.Now()
	log.Println("-----------------------------------------")
	log.Printf("time: %v \n", end.Sub(start))
	log.Printf("assets: %d \n", len(assets))
	log.Printf("requests lost: %d \n", totalLost)
	log.Printf("requests got: %d \n", totalGot)
}

func getAllAssets(it Iterator) []*assetpb.Asset {
	results := make([]*assetpb.Asset, 0)
	for {
		response, err := it.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			return nil
		}
		results = append(results, response)
	}
	return results
}

```
1. this is roughly the same code from the assets inventory provider that calls `ListAssets`
3. you can change the `Parent` param to a different organization. make sure to re-run `gcloud auth login` and `gcloud auth application-default login`
4. the script limits  requests to 100 per minute and retries failed ones. (you can adjust the quota [here](https://console.cloud.google.com/iam-admin/quota))
6. it simulates consumption of 10x the quota 
7. eventually (+10mins) returns `requests got: 1000`, which means it doesn't lose any requests. 
8. before this PR, the same usage would error after 100 requests and then just log the error and continue. you can remove the `rl.Wait()` and `RetryOnResourceExhausted` to see verify

</details>

### Related Issues
- part of https://github.com/elastic/cloudbeat/issues/2054
- closes https://github.com/elastic/cloudbeat/issues/2073

### CODE REVIEW

File: internal/flavors/benchmark/gcp.go [hunk 1]
```diff
@@ -49,6 +49,7 @@ func (g *GCP) NewBenchmark(ctx context.Context, log *logp.Logger, cfg *config.Co
 
 	return builder.New(
 		builder.WithBenchmarkDataProvider(bdp),
+		builder.WithManagerTimeout(cfg.Period),
```
[the remark this answers is not in the document]
Comment (uri-weisman) [reply internal/flavors/benchmark/gcp.go:52]: It makes sense.
Once we update the rest of the cloud providers to have a rate limiter we should consider removing the manager timeout option and configuring all to be limited to the interval period (24h).

File: internal/resources/providers/gcplib/inventory/grpc_rate_limiter.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+})
+
+type AssetsInventoryRateLimiter struct {
+	methods map[string]*rate.Limiter
+	log     *logp.Logger
+}
+
+// a map of asset inventory client methods and their quotas.
+// see https://cloud.google.com/asset-inventory/docs/quota
+var methods = map[string]*rate.Limiter{
```
[the remark this answers is not in the document]
Comment (uri-weisman) [reply internal/resources/providers/gcplib/inventory/grpc_rate_limiter.go:47]: so what happens in case we call a method that is not `ListAssets`? the interceptor is still active?

[the remark this answers is not in the document]
Comment (uri-weisman) [reply internal/resources/providers/gcplib/inventory/grpc_rate_limiter.go:47]: the interceptor will be called but we'll not wait, right?

File: internal/resources/providers/gcplib/inventory/provider.go [hunk 3]
```diff
@@ -466,35 +463,42 @@ func extendWithECS(ctx context.Context, crm *ResourceManagerWrapper, cache map[s
...
-		assets := getAllAssets(p.log, p.inventory.ListAssets(ctx, &assetpb.ListAssetsRequest{
+
+		assets := p.getAllAssets(ctx, &assetpb.ListAssetsRequest{
 			ContentType: assetpb.ContentType_IAM_POLICY,
 			Parent:      parent,
 			AssetTypes:  []string{assetType},
-		}))
-		return extendWithECS(ctx, p.crm, p.crmCache, assets)
+		})
+		extendedAssets := extendWithECS(ctx, p.crm, p.crmCache, assets)
```
Comment (uri-weisman) [inline internal/resources/providers/gcplib/inventory/provider.go:499]: Is there an option of reaching a rate limit with the CRM client as well?
for example, if the customer has many different projects...

If this is an option, we might need to add a limiter or at least a retrier.

**Review by uri-weisman — COMMENTED** [review]
Looks great! 💪 
Left some questions...

File: internal/resources/providers/gcplib/inventory/grpc_rate_limiter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (orouz) [reply internal/resources/providers/gcplib/inventory/grpc_rate_limiter.go:47]: > the interceptor will be called but we'll not wait, right?

yeah the interceptor will just be a pass-through function

File: internal/resources/providers/gcplib/inventory/provider.go (hunk 3 above)
Comment (orouz) [reply internal/resources/providers/gcplib/inventory/provider.go:499]: it looks like the `Get` operation which we use doesn't have a [quota](https://cloud.google.com/resource-manager/docs/limits). the only quotas i see are for `list` and `search`. (v3)

there is however room for improvements here, as `extendWithECS` uses a cache, but the underlining calls to get display names do not. so if for example, `projects/123/orgs/456` gets a cache hit in `extendWithECS` but `projects/789/orgs/456` does not, we still make a call to get the org display name, even though we already have it. 

i'll add caching for display names fetching as well. 

regarding retry - the `Project.Get(..).Context(..).Do()` call doesn't (AFAIK) enable us to retry (`Do()` only gets `key` and `value` back). we could replace the http client the CRM service uses, and account for retry there, but given there doesn't seem to be a quota for our usage (`Get()` only) i think maybe a cache would be sufficient. wdyt?

Comment (uri-weisman) [reply internal/resources/providers/gcplib/inventory/provider.go:499]: yeah, If there is no quota, a thin cache layer will be sufficient

File: internal/resources/providers/gcplib/inventory/map_cache.go
Comment (orouz) [inline internal/resources/providers/gcplib/inventory/map_cache.go:1]: we already have a [cache](https://github.com/elastic/cloudbeat/blob/6554efe33160954d4827a1abd89d9a7a9e281bde/internal/resources/fetching/cycle/cache.go#L45) utility but it is used for single values and is cycle-aware 

this cache just abstracts the repetitive read/write using a plain `map` would require and instead takes a function to get a value which will be used for initial read and assignment.

**Review by uri-weisman — APPROVED** [review]

File: internal/resources/providers/gcplib/inventory/provider.go [hunk 4]
```diff
@@ -444,92 +440,81 @@ func mergeAssetContentType(assets []*assetpb.Asset) []*assetpb.Asset {
...
-		if strings.HasPrefix(parent, "organizations") {
-			assetType = CrmOrgAssetType
-		}
-		assets := getAllAssets(p.log, p.inventory.ListAssets(ctx, &assetpb.ListAssetsRequest{
-			ContentType: assetpb.ContentType_IAM_POLICY,
-			Parent:      parent,
-			AssetTypes:  []string{assetType},
-		}))
-		return extendWithECS(ctx, p.crm, p.crmCache, assets)
+		extendedAssets := ancestorsPoliciesCache.Get(func() []*ExtendedGcpAsset {
```
Comment (jeniawhite) [inline internal/resources/providers/gcplib/inventory/provider.go:478]: You can return this directly without the need to a variable.

File: internal/resources/providers/gcplib/inventory/provider.go [hunk 5]
```diff
@@ -405,19 +400,21 @@ func getAssetsByProject[T any](assets []*ExtendedGcpAsset, log *logp.Logger, f T
...
 	for {
 		response, err := it.Next()
 		if err == iterator.Done {
 			break
 		}
 		if err != nil {
-			log.Errorf("Error fetching GCP Asset: %s", err)
-			return nil
+			p.log.Errorf("Error fetching GCP Asset: %s", err)
+			return results
```
[the remark this answers is not in the document]
Comment (jeniawhite) [reply internal/resources/providers/gcplib/inventory/provider.go:410]: That's great and I can see that eventually this is being appended into assets, we just need to make sure that there is no hidden logic that differentiates between `nil` and actually getting results.
Because we do not want to act is if everything succeeded when we have a partial response.
From what I saw in the code we act in a opportunistic sort of way, but I would consider in the future to recognize when we have partial results and potentially act on it instead of skipping the whole cycle (not in the scope of this PR).

File: internal/resources/providers/gcplib/inventory/provider.go [hunk 6]
```diff
@@ -150,60 +144,62 @@ func (p *ProviderInitializer) Init(ctx context.Context, log *logp.Logger, gcpCon
...
-		config:    gcpConfig,
-		log:       log,
-		inventory: assetsInventoryWrapper,
-		crm:       crmServiceWrapper,
-		crmCache:  make(map[string]*fetching.CloudAccountMetadata),
+		config:                    gcpConfig,
+		log:                       log,
+		inventory:                 assetsInventoryWrapper,
+		crm:                       crmServiceWrapper,
+		cloudAccountMetadataCache: NewMapCache[*fetching.CloudAccountMetadata](),
```
Comment (moukoublen) [inline internal/resources/providers/gcplib/inventory/provider.go:178]: Is `Provider` lifecycle per-cycle or long-lived? I think it is long-lived its initialized once at the startup of cloudbeat, but perhaps I am mistaken here. 

If it is, then shouldn't we be extra careful what we cache once and only? Is cloud account metadata safe to retrieve only once per lifetime? If not we should use one of the many in-mem cache libraries that implements ttl global or per key ttl.

File: internal/flavors/benchmark/gcp.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (moukoublen) [reply internal/flavors/benchmark/gcp.go:52]: @jeniawhite I agree, but this seems to be a new feature we need to create. The upper bound limit could be (without significant effort) something like this: a cycle still running could postpone a maximum of N new cycles and then get canceled. 

However, we should implement it as a new feature and consider the edge scenarios.

[the remark this answers is not in the document]
Comment (orouz) [reply internal/flavors/benchmark/gcp.go:52]: good point. i've opened an issue to handle this scenario - https://github.com/elastic/cloudbeat/issues/2180

File: internal/resources/providers/gcplib/inventory/provider.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (orouz) [reply internal/resources/providers/gcplib/inventory/provider.go:410]: we still log the error like we did before, and the cycle continues without interruption like we did before, it just has a bit more findings to report. overall i think this change is safe, as returning `nil` or an empty `[]*assetpb.Asset` is the same in the sense that we never differentiate between the two, we just don't iterate on an empty value.

File: internal/resources/providers/gcplib/inventory/provider.go [hunk 7]
```diff
@@ -444,92 +440,81 @@ func mergeAssetContentType(assets []*assetpb.Asset) []*assetpb.Asset {
...
-		orgName = crm.getOrganizationDisplayName(ctx, keys.parentOrg)
+		orgName = p.crm.getOrganizationDisplayName(ctx, fmt.Sprintf("organizations/%s", orgId))
 		wg.Done()
 	}()
 	wg.Add(1)
 	go func() {
-		projectName = crm.getProjectDisplayName(ctx, keys.parentProject)
+		// some assets are not associated with a project
+		if projectId != "" {
+			projectName = p.crm.getProjectDisplayName(ctx, fmt.Sprintf("projects/%s", projectId))
```
[the remark this answers is not in the document]
Comment (orouz) [reply internal/resources/providers/gcplib/inventory/provider.go:509]: this does change the behavior we had before, but i think it's ok. we used to try to fetch project names using a project id of empty string, which resulted in an empty project name. we did this multiple times, so sometimes we got the empty project name from cache. in any case, after https://github.com/elastic/cloudbeat/pull/2085 was merged, we don't send empty project names anyway:
https://github.com/elastic/cloudbeat/blob/699424baaad328aaea3335b8baa0934bba5919d5/internal/dataprovider/providers/cloud/data_provider.go#L65

so the outcome of this change is ultimately just not sending redundant api calls to fetch empty project names.

File: internal/resources/providers/gcplib/inventory/provider.go (hunk 6 above)
Comment (orouz) [reply internal/resources/providers/gcplib/inventory/provider.go:178]: before this PR, the cache was a plain map:
```diff
- crmCache map[string]*fetching.CloudAccountMetadata
+ cloudAccountMetadataCache *MapCache[*fetching.CloudAccountMetadata]
```

so it's still the same behavior as before, only now a little less repetitive as `MapCache` just takes away the operations we'd need to do to cache values using a plain `map[string]` 

in general though, i agree this behaviour is probably not correct, even though project/org names probably rarely change, we'd probably still want to get fresh values. i've opened an [issue](https://github.com/elastic/cloudbeat/issues/2182) to address this in a separate PR (as the behaviour itself didn't change from before and is unrelated to this PR)
